keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/MultiplicationVertex.java

```diff
@@ -37,12 +37,12 @@ protected DoubleTensor op(DoubleTensor l, DoubleTensor r) {
     @Override
     protected PartialDerivative forwardModeAutoDifferentiation(PartialDerivative dLeftWrtInput, PartialDerivative dRightWrtInput) {
 
-        PartialDerivative fromLeft = AutoDiffBroadcast.correctForScalarPartialForward(dLeftWrtInput, left.getShape(), this.getShape());
-        PartialDerivative fromRight = AutoDiffBroadcast.correctForScalarPartialForward(dRightWrtInput, right.getShape(), this.getShape());
+        PartialDerivative fromLeft = AutoDiffBroadcast.correctForBroadcastPartialForward(dLeftWrtInput, left.getShape(), this.getShape());
```

Comment:
This code is pretty boilerplate (both forwards and backwards).  It would be nice if we didn't have to put it in loads of the vertices.  I know we've just changed this, but maybe it is best in the differentiators after all?

Reply:
I think it's the operations responsibility to provide the autodiff calculation. This correct for broadcast boilerplate will exist only in operations that allow the broadcast feature. Correcting for it later is really complicated and ends up complicating even the simple element-wise operations. That being said, boilerplate is ugly. Since it only happens in the 4 big arithmetic ops it's not something that will be repeated everywhere and cleaning it up might end up being more complicated than these 2 lines in each op that supports broadcast.